nano-sync/src/circuits/mnt4/pk_tree_5.rs

```diff
@@ -0,0 +1,314 @@
+        )?;
+
+        let position_var =
+            UInt8::alloc_input_vec(cs.ns(|| "alloc position"), self.position.as_ref())?
+                .pop()
+                .unwrap();
+
+        // Process public inputs.
+        next_cost_analysis!(cs, cost, || { "Process public inputs" });
+
+        let chunk_start = VALIDATOR_SLOTS / PK_TREE_BREADTH * (self.position[0] as usize);
+        let chunk_end = VALIDATOR_SLOTS / PK_TREE_BREADTH * (1 + self.position[0] as usize);
+
+        let prepare_signer_bitmap_var =
+            prepare_signer_bitmap_var.to_bits(cs.ns(|| "prepare signer bitmap to bits"))?;
+
+        let prepare_signer_bitmap_var = prepare_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let commit_signer_bitmap_var =
+            commit_signer_bitmap_var.to_bits(cs.ns(|| "commit signer bitmap to bits"))?;
+
+        let commit_signer_bitmap_var = commit_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let mut path_var = position_var.into_bits_le();
+
+        path_var.truncate(PK_TREE_DEPTH);
+
+        // Verify the Merkle proof for the public keys.
+        next_cost_analysis!(cs, cost, || { "Verify Merkle proof pks" });
+
+        let mut bits: Vec<Boolean> = vec![];
+
+        for i in 0..self.pks.len() {
+            bits.extend(SerializeGadget::serialize_g2(
+                cs.ns(|| format!("serialize pks: {}", i)),
+                &pks_var[i],
+            )?);
+        }
+
+        MerkleTreeGadget::verify(
+            cs.ns(|| "verify Merkle proof for pks"),
+            &bits,
+            &pks_nodes_var,
+            &path_var,
+            &pk_commitment_var,
+            &pedersen_generators_var,
+            &sum_generator_g1_var,
+        )?;
+
+        // Verifying prepare aggregate public key commitment. It just checks that the prepare
+        // aggregate public key given as private input is correct by committing to it and comparing
+        // the result with the prepare aggregate public key commitment given as a public input.
+        next_cost_analysis!(cs, cost, || { "Verify prepare agg pk commitment" });
+
+        let prepare_agg_pk_bits = SerializeGadget::serialize_g2(
+            cs.ns(|| "serialize prepare agg pk"),
+            &prepare_agg_pk_var,
+        )?;
+
+        let pedersen_commitment = PedersenCommitmentGadget::evaluate(
+            cs.ns(|| "reference prepare agg pk commitment"),
+            &prepare_agg_pk_bits,
+            &pedersen_generators_var,
+            &sum_generator_g1_var,
+        )?;
+
+        let pedersen_bits = SerializeGadget::serialize_g1(
+            cs.ns(|| "serialize prepare pedersen commitment"),
+            &pedersen_commitment,
+        )?;
+
+        let pedersen_bits = reverse_inner_byte_order(&pedersen_bits[..]);
+
+        let mut reference_commitment = Vec::new();
+
+        for i in 0..pedersen_bits.len() / 8 {
+            reference_commitment.push(UInt8::from_bits_le(&pedersen_bits[i * 8..(i + 1) * 8]));
+        }
```

Comment:
Add comment what you're doing here.
If I'm not mistaken you convert the `pedersen_bits` into a `Vec<UInt8>`.

Reply:
Seems a bit unnecessary, but I added it.